language_service/src/completion.rs

```diff
@@ -1,809 +1,334 @@
-    let source_offset = (package_offset - source.offset)
-        .try_into()
-        .expect("offset can't be converted to uszie");
-    let before_offset = &source.contents[..source_offset];
-    let mut indent = match before_offset.rfind(|c| c == '{' || c == '\n') {
-        Some(begin) => {
-            let indent = &before_offset[begin..];
-            indent.strip_prefix('{').unwrap_or(indent)
-        }
-        None => before_offset,
-    }
-    .to_string();
-    if !indent.starts_with('\n') {
-        indent.insert(0, '\n');
+        .relative_sources()
+        .find(|s| s.offset == source.offset)
+        .expect("source should exist in the user source map")
+        .name;
+
+    if log_enabled!(Trace) {
+        let last_char = if source_offset > 0 {
+            source.contents[(package_offset as usize - 1)..]
+                .chars()
+                .next()
+        } else {
+            None
+        };
+        trace!("the character before the cursor is: {last_char:?}");
     }
-    indent
-}
 
-struct CompletionListBuilder {
-    current_sort_group: SortPriority,
-    items: FxHashSet<CompletionItem>,
+    // What kinds of words are expected at the cursor location?
+    let expected_words_at_cursor = expected_word_kinds(
+        compilation,
+        &source_name_relative,
+        &source.contents,
+        source_offset,
+    );
+
+    // Now that we have the information from the parser about what kinds of
+    // words are expected, gather the actual words (identifiers, keywords, etc) for each kind.
+
+    // Keywords and other hardcoded words
+    let hardcoded_completions = collect_hardcoded_words(expected_words_at_cursor);
+
+    // The tricky bit: globals, locals, names we need to gather from the compilation.
+    let name_completions = collect_names(
+        expected_words_at_cursor,
+        package_offset,
+        compilation,
+        position_encoding,
+    );
+
+    // We have all the data, put everything into a completion list.
+    into_completion_list(once(hardcoded_completions).chain(name_completions))
 }
 
-impl CompletionListBuilder {
-    fn new() -> Self {
-        CompletionListBuilder {
-            current_sort_group: 1,
-            items: FxHashSet::default(),
-        }
+/// Invokes the parser to determine what kinds of words are expected at the cursor location.
+fn expected_word_kinds(
+    compilation: &Compilation,
+    source_name_relative: &str,
+    source_contents: &str,
+    cursor_offset: u32,
+) -> WordKinds {
+    match &compilation.kind {
+        CompilationKind::OpenProject {
+            package_graph_sources,
+        } => possible_words_at_offset_in_source(
+            source_contents,
+            Some(source_name_relative),
+            package_graph_sources.root.language_features,
+            cursor_offset,
+        ),
+        CompilationKind::Notebook { project } => possible_words_at_offset_in_fragments(
+            source_contents,
```

Comment:
How do source contents work for notebook cells? I was curious about this and tried playing around with it, ended up seeing some behaviors I didn't expect, so wanted to confirm whether there are known bugs/limitations here or if this is something that bears investigating. I'll follow up with screen shots and more explanation...

Reply:
When I try out the language service with these changes, completions in a file work as I'd expect. I get completions for local variables that are in-scope at my current line (the `x` from the line above):
<img width="666" alt="image" src="https://github.com/user-attachments/assets/53806d80-5742-4240-aef0-9e01dae5b7fe">
or for locally defined callables:
<img width="666" alt="image" src="https://github.com/user-attachments/assets/165e92e2-a31c-4ea2-8ead-303add27cd82">

But in notebook cells, I don't see these elements if they come from top-level statements (it does work as expected when inside a callable context). So the local variables don't show `x` or `IntOne` from the lines above my cursor:
<img width="666" alt="image" src="https://github.com/user-attachments/assets/a528e6d5-76aa-4019-a0fe-af5b5c67ec7b">
<img width="650" alt="image" src="https://github.com/user-attachments/assets/e6bf4895-c8d0-4b5f-8493-985ac5d75ac7">

Oddly enough, the locals and functions DO show up under two different circumstances. If instead of triggering completions with ctrl+space I just start typing, they show up:
<img width="650" alt="image" src="https://github.com/user-attachments/assets/e76c8c29-7a54-40ef-ad2f-c7ca2e9c18ee">
<img width="650" alt="image" src="https://github.com/user-attachments/assets/812f9510-e893-4ea6-849c-e052b7642a8b">
OR if I put the definitions in a previous cell, then they show up even if I trigger completions before I start typing:
<img width="650" alt="image" src="https://github.com/user-attachments/assets/82f547a2-30d7-4fed-8861-ca892b7227d5">

I don't think this is a major issue and would probably be fine as a follow up investigation/bug-fix, but it does tell me that there is some difference between the response of completions when triggered manually vs when typing specifically for notebook cells that I don't really understand.

Reply:
This turned out to be a preexisting bug. I'll investigate.

Reply:
Completions were a distraction, actually - having a broken expression at the very end of the cell breaks the parsing of the entire cell, there's no AST at all! Notice none of the editor features (hover, go to def etc) work for _any_ code above that expression.

I'll just file a separate bug for this.

Reply:
Tracked in #1960

Reply:
Oh wow, I didn't realize that. Thanks for investigating!